feat(worker): support workers to run natively on windows

alpetric opened the pull request:
The goal was to be able to build and run workers on windows without the need of docker or WSL.

As a first step the worker should support:
  - python
  - powershell
  - typescript (bun)

Trying to keep code changes to a minimum but includes some minor automatic `rustfmt`

Notes:
- [on building from source on windows](https://gist.github.com/alpetric/9bb2299d2c65a08ca23f9ecdc468e268)
- [on running the workers with a pre-built executable (aka .exe) on windows](https://gist.github.com/alpetric/07f18f7e505f90bfe2c4859a8bcb47d1)
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Add Windows support for workers by updating executors and environment configurations for compatibility.
> 
>   - **Windows Support**:
>     - Add Windows-specific configurations in `bash_executor.rs`, `bun_executor.rs`, `go_executor.rs`, `python_executor.rs`, and `rust_executor.rs`.
>     - Use `cfg(windows)` to conditionally compile Windows-specific code.
>     - Set `SystemRoot` and other environment variables for Windows in executors.
>   - **Executors**:
>     - Modify `handle_powershell_job()` in `bash_executor.rs` to support Windows paths and environment.
>     - Update `handle_bun_job()` in `bun_executor.rs` to handle Windows-specific paths and environment variables.
>     - Adjust `handle_python_job()` in `python_executor.rs` for Windows compatibility.
>     - Update `handle_rust_job()` in `rust_executor.rs` to set Windows-specific environment variables.
>   - **Miscellaneous**:
>     - Minor formatting changes and cleanup in various files.
>     - Add `SYSTEM_ROOT` constant for Windows in `worker.rs`.
>     - Introduce `build_windows_worker.yml` GitHub Actions workflow for building Windows workers.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=windmill-labs%2Fwindmill&utm_source=github&utm_medium=referral)<sup> for fdf0b0177c4484ae692fd88ef389664feffc8562. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

alpetric commented on backend/windmill-worker/src/bash_executor.rs:
```diff
@@ -292,6 +299,17 @@ Get-Module -ListAvailable | Import-Module
+    let profile = format!(
+        "#$PSModuleAutoloadingPreference = 'None'
+$PSModulePathBackup = $env:PSModulePath
+#$env:PSModulePath = \"C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Modules;C:\\Program Files\\PowerShell\\7\\Modules\"
```
@HugoCasa not quite sure what's going on here, specifically 

```
$env:PSModulePath = ($Env:PSModulePath -split ':')[-1]
Get-Module -ListAvailable | Import-Module	Get-Module -ListAvailable | Import-Module
```
Why do we pick the last element of the list of module paths to import? With powershell 7 in a clean windows install there are a bunch of other paths one could import modules from.

HugoCasa replied:
best way to do that for both unix and windows would be to use $PSHOME/Modules instead of last path in PSModulePath.
The goal here is to only preload the default modules.
source: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7.4

HugoCasa replied:
@alpetric  https://github.com/windmill-labs/windmill/pull/4448

alpetric commented on backend/windmill-worker/src/go_executor.rs:
```diff
@@ -225,12 +225,20 @@ func Run(req Req) (interface{{}}, error){{
         })?;
 
+        #[cfg(windows)]
+        std::os::windows::fs::symlink_dir(&bin_path, &target).map_err(|e| {
```
it's a folder

rubenfiszel commented on backend/windmill-worker/src/python_executor.rs:
```diff
@@ -430,6 +433,13 @@ mount {{
                 ),
         )?;
     } else {
+        #[cfg(windows)]
+        reserved_variables.insert(
+            "PYTHONPATH".to_string(),
+            additional_python_paths_folders.replace(":", ";"),
+        );
+
+        #[cfg(unix)]
         reserved_variables.insert("PYTHONPATH".to_string(), additional_python_paths_folders);
```
less bloat by just shadowing additional_python_paths_folders on windows:
```
#[cfg(windows)]
let additional_python_paths_folders = additional_python_paths_folders.replace(":", ";");
```

rubenfiszel commented on backend/windmill-worker/src/bash_executor.rs:
```diff
@@ -325,52 +364,86 @@ $env:PSModulePath = \"{}:$PSModulePathBackup\"",
+    #[cfg(windows)]
+    {
+        let mut cmd_args = vec![r".\wrapper.ps1".to_string()];
+        cmd_args.extend(pwsh_args.iter().map(|x| x.replace("--", "-")));
+
+        child = Command::new(POWERSHELL_PATH.as_str())
             .current_dir(job_dir)
             .env_clear()
             .envs(envs)
             .envs(reserved_variables)
-            .env("TZ", TZ_ENV.as_str())
             .env("PATH", PATH_ENV.as_str())
             .env("BASE_INTERNAL_URL", base_internal_url)
             .env("HOME", HOME_ENV.as_str())
+            .env("SystemRoot", SYSTEM_ROOT.as_str())
+            .env(
+                "TMP",
+                std::env::var("TMP").unwrap_or_else(|_| String::from("/tmp")),
+            )
+            .env(
+                "PATHEXT",
+                ".COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL",
+            )
             .args(cmd_args)
             .stdout(Stdio::piped())
             .stderr(Stdio::piped())
-            .spawn()?
-    };
+            .spawn()?;
+    }
```
is there a way we can factorize what is common between the 2 more elegantly

rubenfiszel commented on backend/windmill-worker/src/rust_executor.rs:
```diff
@@ -279,12 +314,20 @@ pub async fn handle_rust_job(
             ))
         })?;
 
+        #[cfg(windows)]
```
can be factorized

rubenfiszel commented on backend/windmill-worker/src/rust_executor.rs:
```diff
@@ -173,9 +197,20 @@ pub async fn build_rust_crate(
         .stdout(Stdio::piped())
         .stderr(Stdio::piped());
+
+    build_rust_cmd.args(vec!["build", "--release"]);
```
looks like it could be kept above but not important

rubenfiszel commented on backend/windmill-worker/src/rust_executor.rs:
```diff
@@ -23,14 +23,30 @@ use crate::{
     static ref CARGO_PATH: String = format!("{}/bin/cargo", std::env::var("CARGO_HOME").unwrap_or("/usr/local/cargo/bin/cargo".to_string()));
 }
 
+#[cfg(windows)]
+lazy_static::lazy_static! {
+    static ref HOME_DIR: String = std::env::var("HOME").expect("Could not find the HOME environment variable");
+    static ref CARGO_HOME: String = std::env::var("CARGO_HOME").unwrap_or_else(|_| {
+        format!("{}\\.cargo", *HOME_DIR)
+    });
+    static ref RUSTUP_HOME: String = std::env::var("RUSTUP_HOME").unwrap_or_else(|_| {
+        format!("{}\\.rustup", *HOME_DIR)
+    });
+    static ref CARGO_PATH: String = format!("{}/bin/cargo.exe", *CARGO_HOME);
+}
```
if only the default change then assign the unwrap_or_else to a const  default and make that be different
=>

can be factorized

rubenfiszel commented on backend/windmill-worker/src/go_executor.rs:
```diff
@@ -225,12 +225,20 @@ func Run(req Req) (interface{{}}, error){{
         })?;
 
+        #[cfg(windows)]
+        std::os::windows::fs::symlink_dir(&bin_path, &target).map_err(|e| {
```
same map_err, can be factorized

rubenfiszel commented on backend/windmill-worker/src/bun_executor.rs:
```diff
@@ -815,14 +836,25 @@ pub async fn handle_bun_job(
+        #[cfg(windows)]
+        let target = format!("{job_dir}\\main.js");
+        #[cfg(windows)]
+        std::os::windows::fs::symlink_dir(&local_path, &target).map_err(|e| {
```
same map_err can be factorized

rubenfiszel commented on backend/windmill-worker/src/bun_executor.rs:
```diff
@@ -540,6 +557,7 @@ pub async fn pull_codebase(w_id: &str, id: &str, job_dir: &str) -> Result<()> {
         if is_tar {
             extract_tar(fs::read(bun_cache_path)?.into(), job_dir).await?;
         } else {
+            #[cfg(unix)]
```
missing a cp on windows

rubenfiszel commented on backend/windmill-worker/src/bun_executor.rs:
```diff
@@ -553,6 +571,7 @@ pub async fn pull_codebase(w_id: &str, id: &str, job_dir: &str) -> Result<()> {
         if is_tar {
             extract_tar(bytes, job_dir).await?;
         } else {
+            #[cfg(unix)]
             tokio::fs::symlink(bun_cache_path, dst).await?;
```
missing a cp on windows

alpetric commented:
@rubenfiszel addressed your comments and carefully reviewed impact on windows. The biggest risk are the simplifications of the RUST env vars but they look good but (most of) the rest is behind feature flags.

also we can merge this PR first https://github.com/windmill-labs/windmill/pull/4448
